webserver/templates/user/scraper.js

```diff
@@ -1,256 +1,296 @@
-    // less than 10 pages waiting to submit
-    while (page <= numberOfPages && activeSubmissions < 10) {
-        page += 1;
-        activeSubmissions++;
-        getLastFMPage(page);
+
+    function encodeScrobbles(root) {
+        var scrobbles = root.getElementsByClassName("js-link-block");
+        var parsedScrobbles = map(function(rawScrobble) {
+            var scrobble = new Scrobble(rawScrobble);
+            return scrobble.asJSONSerializable();
+        }, scrobbles);
+
+        var structure = {
+            "listen_type" : "import",
+            "payload"     : parsedScrobbles
+        };
+
+        return structure;
     }
-}
-
-function pageDone() {
-    activeSubmissions--;
-    numCompleted++;
-    getNextPagesIfSlots();
-}
-
-function updateMessage(message) {
-    document.getElementById("listen-progress-container").innerHTML =  "" +
-        "<img src='{{ url_for('static', filename='img/listenbrainz-logo.svg', _external=True) }}' height='75'><br><br>" +
-        message + "<br>" +
-        "<span style='display:none'>" +
-        "reportedScrobbles " + timesReportScrobbles +
-        ", getPage " + timesGetPage + ", number4xx " + times4Error +
-        ", number5xx " + times5Error + ", page " + page + "</span>" +
-        "<br><span style='font-size:6pt; position:absolute; bottom:1px; right: 3px'>v"+version+"</span>";
-}
-
-document.body.insertAdjacentHTML( 'afterbegin', '<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">');
-document.body.insertAdjacentHTML( 'afterbegin', '<div style="position:absolute; top:200px; z-index: 200000000000000; width:500px; margin-left:-250px; left:50%; background-color:#fff; box-shadow: 0 19px 38px rgba(0,0,0,0.30), 0 15px 12px rgba(0,0,0,0.22); text-align:center; padding:50px;" id="listen-progress-container"></div>');
-updateMessage("");
-getNextPagesIfSlots();
+
+    function ListenBrainzSubmissionQueue(){
+        var self = this;
+        self.remain = 1;
+        self.timeout = 0;
+        self.queue = [];
+        self.processingReq = false;
+        self.count = 0;
+        self.doneSubmissions = false;
+        self.start = true;
+
+        self.eventHandlers = {};
+
+        self.on = function(eventName, eventHandler){
+            self.eventHandlers[eventName] = eventHandler;
+        };
+
+        self.push = function(struct){
+            if(self.start){
+                if (self.eventHandlers.update){
+                    self.eventHandlers.update(1);
+                }
+                self.start = false;
+            }
+            self.queue.push(struct);
+            if(!self.processingReq){
+                self.processReq();
+            }
+        };
+
+        self.done = function(){
+            self.doneSubmissions = true;
+        };
+
+        self.processReq = function(){
+            self.processingReq = true;
+            struct = self.queue.pop();
```

Comment:
Should be `var struct =`. Right now it's setting a global, so `setTimeout` calls that haven't run yet are reading a different value (possibly `undefined` if there are no items left).